Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,468 @@
+    conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromBlocks_map,
+    fromBlocks_mul_fromRows, fromColumns_mul_fromRows]
+  simp only [map_zero, Matrix.map_zero, Matrix.zero_mul, add_zero, Matrix.mul_zero]
+  rw [Matrix.mul_assoc]
+  rw [map_zero]
+
+open scoped ComplexOrder
+lemma eig_vals_ne_zero_pos {m n: Type} [Fintype m] [Fintype n] [DecidableEq n]
+    (A : Matrix m n 𝕂) (z: {a // (isHermitian_transpose_mul_self A).eigenvalues a ≠ 0 }) :
+    0 < ((isHermitian_transpose_mul_self A).eigenvalues z) :=
+  lt_of_le_of_ne
+    (Matrix.PosSemidef.eigenvalues_nonneg (posSemidef_conjTranspose_mul_self _) _) -- 0 ≤ _
+    (z.prop.symm) -- 0 ≠ _
+
+lemma svdσ_inv (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdσ⁻¹ =
+    (reindex
+      (finRankEquivEigsConjTransposeMulSelf A).symm
+      (finRankEquivEigsConjTransposeMulSelf A).symm)
+      (diagonal (fun (i : {a // (isHermitian_transpose_mul_self A).eigenvalues a ≠ 0}) =>
+        1 / Real.sqrt ((isHermitian_transpose_mul_self A).eigenvalues i))) := by
+  apply inv_eq_right_inv
+  rw [svdσ]
+  simp only [ne_eq, reindex_apply, submatrix_diagonal_equiv, diagonal_mul_diagonal,
+    Function.comp_apply]
+  rw [← diagonal_one, diagonal_eq_diagonal_iff]
+  intros i
+  rw [mul_one_div_cancel]
+  apply ne_of_gt
+  apply (Real.sqrt_pos.2 (eig_vals_ne_zero_pos _ _))
+
+lemma σ_inv_μ_σ_inv_eq_one (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (A.svdσ⁻¹)ᴴ  *  A.svdμ  *  A.svdσ⁻¹ = 1 := by
+  rw [svdσ_inv, svdμ]
+  simp only [ne_eq, one_div, reindex_apply, submatrix_diagonal_equiv, diagonal_conjTranspose,
+    star_trivial, diagonal_mul_diagonal, Function.comp_apply]
+  rw [← diagonal_one, diagonal_eq_diagonal_iff]
+  intro i
+  rw [mul_comm, ← mul_assoc, ← mul_inv, Real.mul_self_sqrt, inv_mul_cancel]
+  apply ne_of_gt (eig_vals_ne_zero_pos A _)
+  apply le_of_lt (eig_vals_ne_zero_pos A _)
+
+lemma IsUnit_det_svdσ (A : Matrix (Fin M) (Fin N) 𝕂) : IsUnit (A.svdσ.det) := by
+  unfold svdσ
+  rw [reindex_apply]
+  simp only [ne_eq, submatrix_diagonal_equiv, det_diagonal, Function.comp_apply]
+  apply Ne.isUnit (Finset.prod_ne_zero_iff.2
+    (fun i _ => (ne_of_gt (Real.sqrt_pos.2 (eig_vals_ne_zero_pos _ _)))))
+
+lemma IsUnit_det_svdσ_mapK (A : Matrix (Fin M) (Fin N) 𝕂) :
+    IsUnit (det (map A.svdσ (algebraMap ℝ 𝕂))) := by
+  unfold svdσ
+  simp only [ne_eq, reindex_apply, submatrix_diagonal_equiv, map_zero, diagonal_map,
+    Function.comp_apply, det_diagonal]
+  rw [isUnit_iff_ne_zero, Finset.prod_ne_zero_iff]
+  intro i
+  simp only [Finset.mem_univ, ne_eq, map_eq_zero, forall_true_left]
+  apply (ne_of_gt (Real.sqrt_pos.2 (eig_vals_ne_zero_pos _ _)))
+
+lemma svdσ_inv_mapK (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (map (A.svdσ) (algebraMap ℝ 𝕂))⁻¹ = (map (A.svdσ)⁻¹ (algebraMap ℝ 𝕂)) := by
+  apply inv_eq_left_inv
+  rw [← map_mul, nonsing_inv_mul]
+  simp only [map_zero, _root_.map_one, map_one]
+  apply IsUnit_det_svdσ
+
+lemma U₁_conjTranspose_mul_U₁ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdU₁ᴴ  *  A.svdU₁ = 1 := by
```

Comment:
I see that a lot of these multiplications are double-spaced.
```suggestion
lemma U₁_conjTranspose_mul_U₁ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdU₁ᴴ * A.svdU₁ = 1 := by
```
I think that with a regex you can search for double spaces that are preceded by a non-space. Please make them single spaced.